### ISSUE TITLE
Make memorise in FSharpLu.Json thread-safe

### ISSUE THREAD

**Issue by cmeeren** [Microsoft/fsharplu#36]
I'm occasionally receiving these exceptions in a production API:

```
2018-03-27 10:59:50.500 +02:00 [Error] Unhandled exception while executing request
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at <StartupCode$Microsoft-FSharpLu-Json>.$Compact.-ctor@69.Invoke(Type key)
[rest omitted]
```

It seems that there is a `NullReferenceException` on [this line](https://github.com/Microsoft/fsharplu/blob/master/FSharpLu.Json/Compact.fs#L28). A quick web search turned up e.g. [this SO answer](https://stackoverflow.com/a/2195287/2978652), which suggests that such exceptions may be due to threading issues.

I suggest you make `memorise` thread-safe, by using ConcurrentDictionary or locking.

**Comment by blumu** [comment]
Good catch @cmeeren. We should have caught this bug when reviewing the pull request. Here is a related blog post: http://fsharpnews.blogspot.com/2013/10/a-thread-safe-object-cache.html?m=1

**Comment by Rangoric** [comment]
I get this while running xUnit unit tests using the Json Compact Serializer.

Is there a work around?
Is this something I should try to do a PR for (still a newbie for F#, not new to .Net)?
Is there anything I can do at all to help with this? I have a minimal reproduction of it that fails just about everytime on OSX, that I could email. It seems to be from having many tests doing serialization/deserialization.

**Comment by cmeeren** [comment]
I would also appreciate a fix for this, since it's the most common exception from my web API. I can try a simple PR if I can get some pointers on whether you want to use ConcurrentDictionary or locking, but AFAIK this should be a simple five minute fix when done from your end. :)

**Comment by blumu** [comment]
@cmeeren this bug is already fixed in the latest release. Closing it.

**Comment by cmeeren** [comment]
Brilliant, thanks! :)

**Comment by cmeeren** [comment]
Have you published to NuGet? It says "last updated 2 months ago" https://www.nuget.org/packages/Microsoft.FSharpLu.Json/

**Comment by blumu** [comment]
@cmeeren Just published 0.10.21: https://www.nuget.org/packages/Microsoft.FSharpLu.Json/0.10.21
